ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/CommandDeduplicationBase.scala

```diff
@@ -0,0 +1,309 @@
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+private[testtool] abstract class CommandDeduplicationBase(
+    timeoutScaleFactor: Double,
+    ledgerTimeInterval: FiniteDuration,
+) extends LedgerTestSuite {
+
+  val deduplicationTime: FiniteDuration = 3.seconds * timeoutScaleFactor match {
+    case duration: FiniteDuration => duration
+    case _ =>
+      throw new IllegalArgumentException(s"Invalid timeout scale factor: $timeoutScaleFactor")
+  }
+  val defaultDeduplicationWindowWait: FiniteDuration = deduplicationTime + ledgerTimeInterval * 2
+
+  def runGivenDeduplicationWait(context: ParticipantTestContext)(test: Duration => Future[Unit])(
+      implicit ec: ExecutionContext
+  ): Future[Unit]
+
+  def testNamingPrefix: String
+
+  test(
+    s"${testNamingPrefix}SimpleDeduplicationBasic",
+    "Deduplicate commands within the deduplication time window",
+    allocate(SingleParty),
+    runConcurrently = false,
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    lazy val requestA1 = ledger
+      .submitRequest(party, DummyWithAnnotation(party, "First submission").create.command)
+      .update(
+        _.commands.deduplicationPeriod := DeduplicationPeriod.DeduplicationTime(
+          deduplicationTime.asProtobuf
+        )
+      )
+    lazy val requestA2 = ledger
+      .submitRequest(party, DummyWithAnnotation(party, "Second submission").create.command)
+      .update(
+        _.commands.deduplicationPeriod := DeduplicationPeriod
+          .DeduplicationDuration(
+            deduplicationTime.asProtobuf
+          ), //same semantics as `DeduplicationTime`
+        _.commands.commandId := requestA1.commands.get.commandId,
+      )
+    runGivenDeduplicationWait(ledger) { deduplicationWait =>
+      for {
+        // Submit command A (first deduplication window)
+        // Note: the second submit() in this block is deduplicated and thus rejected by the ledger API server,
+        // only one submission is therefore sent to the ledger.
+        ledgerEnd1 <- ledger.currentEnd()
+        _ <- ledger.submit(requestA1)
+        failure1 <- ledger
+          .submit(requestA1)
+          .mustFail("submitting the first request for the second time")
+        completions1 <- ledger.firstCompletions(ledger.completionStreamRequest(ledgerEnd1)(party))
+        // Wait until the end of first deduplication window
+        _ <- Delayed.by(deduplicationWait)(())
+
+        // Submit command A (second deduplication window)
+        // Note: the deduplication window is guaranteed to have passed on both
+        // the ledger API server and the ledger itself, since the test waited more than
+        // `deduplicationSeconds` after receiving the first command *completion*.
+        // The first submit() in this block should therefore lead to an accepted transaction.
+        ledgerEnd2 <- ledger.currentEnd()
+        _ <- ledger.submit(requestA2)
```

Comment:
This reasoning is not correct. The underlying ledger is allowed to extend the deduplication period as it sees fit. So it should be correct if the first command fails with "already exists".

Reply:
Indeed, so we could simplify our tests and just validate that we get an already exists for a duplicate command and not validate the expiration of the deduplication window.  
But having knowledge about the ledger implementation I think it's a valid point to validate the expiration of the deduplication window as well.

Reply:
Yes. Let's discuss how we go about these differences in person.